tests/hypothesis_strategies.py

```diff
@@ -0,0 +1,158 @@
+
+categories = st.lists(st.text(min_size=1), min_size=2, unique=True)
+"""A strategy that generates parameter categories."""
+
+
+@st.composite
+def smiles(draw: st.DrawFn):
+    """Generate short SMILES strings."""
+    n_atoms = draw(st.integers(min_value=0, max_value=19))
+    string = "C"
+    for _ in range(n_atoms):
+        next_atom = random.choice("CNO") if string[-1] == "C" else random.choice("C")
+        string += next_atom
+    return string
+
+
+@st.composite
+def substance_data(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.substance.SubstanceParameter`."""
+    names = draw(st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True))
+    substances = draw(st.lists(smiles(), min_size=len(names), max_size=len(names)))
+    return dict(zip(names, substances))
+
+
+@st.composite
+def custom_encodings(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.custom.CustomDiscreteParameter`."""
+    index = st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True)
+    cols = columns(
+        names_or_number=10,
+        elements=st.floats(allow_nan=False, allow_infinity=False),
+        unique=True,
+        dtype=DTypeFloatNumpy,
+    )
+    return draw(data_frames(index=index, columns=cols))
+
+
+@st.composite
+def numerical_discrete_parameter(
+    draw: st.DrawFn,
+):
+    """Generate class:`baybe.parameters.numerical.NumericalDiscreteParameter`."""
+    name = draw(parameter_name)
+    values = draw(
+        st.lists(
+            st.one_of(
+                st.integers(),
+                st.floats(allow_infinity=False, allow_nan=False),
+            ),
+            min_size=2,
+            unique=True,
+        )
+    )
+
+    # Reject examples where the tolerance validator cannot be satisfied
+    try:
+        return NumericalDiscreteParameter(name=name, values=values)
+    except NumericalUnderflowError:
+        assume(False)
+
+
+@st.composite
+def numerical_continuous_parameter(draw: st.DrawFn):
+    """Generate class:`baybe.parameters.numerical.NumericalContinuousParameter`."""
+    name = draw(parameter_name)
+    lower = draw(st.floats(max_value=_largest_lower_interval, allow_infinity=False))
+    upper = draw(st.floats(min_value=lower, exclude_min=True, allow_infinity=False))
+    return NumericalContinuousParameter(name=name, bounds=(lower, upper))
+
+
+@st.composite
+def categorical_parameter(draw: st.DrawFn):
+    """Generate class:`baybe.parameters.categorical.CategoricalParameter`."""
+    name = draw(parameter_name)
+    values = draw(categories)
+    encoding = draw(st.sampled_from(CategoricalEncoding))
+    return CategoricalParameter(name=name, values=values, encoding=encoding)
+
+
+@st.composite
+def task_parameter(draw: st.DrawFn):
+    """Generate class:`baybe.parameters.categorical.TaskParameter`."""
+    name = draw(parameter_name)
+    values = draw(categories)
+    active_values = random.sample(values, random.randint(0, len(values)))
```

Comment:
@AVHopp: for random.randint, the upper end included:

@Scienfitz: This one is REALLY strange. First: you are right, should be 1. However, the strange part is that the hypothesis test did not create a failing example. I double checked and, indeed, in none of the runs (no matter how large I set the number of examples to be generated), the lower bound of randint is drawn. Even more surprisingly, it **does** draw the lower bound occasionally when you redraw several times. So something really strange is going on with the RNG here. Replacing randint with the numpy version, for example, fixes the error!?!? Anyway, I also realized that using a hypothesis built-in generation has the additional benefit that it allows for shrinkage, so I've replaced that line of code altogether.

Please double-check both and resolve if you are happy. And do let me know if one of you has an explanation for the above!!

Reply:
Looks good to me. If @Scienfitz agrees we can resolve imo.

Reply:
yep
(the weird effect adrian describes is concerning tho, does anyone understand RNG?)

Reply:
Regarding the weird behavior: I looked a bit deeper and think that it was probably due to wrong use of the framework. Since I manually added a different source of randomness that was not under control of hypothesis, the generated scenarios did not really cover the space of possibilities. I've opened an [issue](https://github.com/HypothesisWorks/hypothesis/issues/3810) on their page just to make sure this is what went wrong. So I'll keep track of it from there and resolve here.

Reply:
@Scienfitz, @AVHopp, short update: turns out the situation is how I expected it: bug was due to wrong usage of the framework (see issue linked above). So everything's good now, will add one more commit that also removes the other non-built in generators.